File: src-tauri/src/fa_preflight.rs
// ---------------------------------------------------------------------------
// Forced-alignment PRE-FLIGHT readiness check (WS1 Session M, Step 4).
//
// WHY THIS EXISTS. Before Session M, the only way to learn that forced
// alignment could not run was to run a full Apply Sync and read the FA-fallback
// entry AFTER it finished — several minutes of Whisper work for a run that was
// never going to use FA in the first place. Worse, the most common failure (the
// bundled onnxruntime runtime not loading) was invisible in the app entirely.
//
// This command answers, up front and cheaply, the three questions the FA gate's
// own `runForcedAlignmentForSync` can only answer by failing: is the runtime
// library loadable, and is the model for this language present. (Capability and
// language RESOLUTION are decided on the TS side, where the project object and
// the `isFaCapable`/`resolveFaLanguage` helpers already live — this command is
// the half that requires the backend: disk and the native runtime.) The TS
// caller folds all four into one pre-flight log entry the user sees before the
// sync starts, and skips the entry when the gate is closed.
//
// COST. The runtime probe is a dlopen + `ort` env init (milliseconds, no
// model). The model check is a path resolution + existence stat — it does NOT
// hash the ~1.2 GiB model (the real run's `verify_model_manifest` still owns
// that). A pre-flight is meant to be run before every FA sync, so it must stay
// this side of cheap.
// ---------------------------------------------------------------------------

use crate::fa::FaError;

/// The backend half of the FA readiness report. `camelCase` for the TS side.
#[derive(serde::Serialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub struct FaPreflightReport {
    /// Whether this binary was compiled with the `fa-inference` feature at all.
    /// `false` in a plain `tauri:dev`/`tauri:build` — FA can never run there
    /// regardless of runtime/model, and the report says so rather than
    /// reporting a runtime failure that is really a build-config fact.
    pub feature_compiled: bool,
    /// The onnxruntime C library resolved AND loaded+init'd successfully.
    pub runtime_ok: bool,
    /// Human-readable: the resolved dylib path on success, or the verbatim
    /// resolution/loader error on failure (the same class of message the
    /// FA-fallback entry now surfaces).
    pub runtime_detail: String,
    /// The FA model for `language` was found on disk (path-resolved + exists).
    /// Not hash-verified here — that is the real run's job.
    pub model_present: bool,
    /// Human-readable: the resolved model path, or the searched-paths error.
    pub model_detail: String,
    /// Echoed back so the log entry and the run can be checked to agree on
    /// which language the readiness was computed for.
    pub language: String,
}

/// Reports FA runtime + model readiness for `language`, without running
/// inference. Never throws for a "not ready" condition — those are reported in
/// the boolean fields; it only `Err`s on a genuinely broken IPC precondition
/// (none today, so it is effectively infallible, but the signature matches the
/// other FA commands so the TS `invoke` shape is uniform).
#[tauri::command]
pub async fn fa_preflight(
    app: tauri::AppHandle,
    language: String,
) -> Result<FaPreflightReport, FaError> {
    // Model presence is a feature-independent fact (the resolver is compiled in
    // both configs), so check it the same way regardless.
    let (model_present, model_detail) = match crate::fa::fa_model_path(&app, &language) {
        Ok(path) => (true, path.to_string_lossy().to_string()),
        Err(e) => (false, e.message),
    };

    #[cfg(feature = "fa-inference")]
    let (feature_compiled, runtime_ok, runtime_detail) = {
        match crate::fa_onnx::probe_ort_runtime(&app) {
            Ok(dylib) => (true, true, format!("onnxruntime loaded from {dylib}")),
            Err(e) => (true, false, e.to_string()),
        }
    };

    #[cfg(not(feature = "fa-inference"))]
    let (feature_compiled, runtime_ok, runtime_detail) = (
        false,
        false,
        "forced-alignment inference is not compiled into this build (the `fa-inference` Cargo \
         feature is off — see CLAUDE.md §2, `npm run tauri:dev:fa`)."
            .to_string(),
    );

    Ok(FaPreflightReport {
        feature_compiled,
        runtime_ok,
        runtime_detail,
        model_present,
        model_detail,
        language,
    })
}

File: src/services/faPreflight.ts
/**
 * @license
 * SPDX-License-Identifier: Apache-2.0
 */

// ---------------------------------------------------------------------------
// Forced-alignment PRE-FLIGHT readiness check — TS side (WS1 Session M, Step 4).
//
// Combines the two halves of "can FA actually run for this project right now":
//   - the FRONTEND facts — runtime capability (`isFaCapable`) and language
//     RESOLUTION (`resolveFaLanguage` + the supported-set check), decided here
//     because the project object and those helpers live on this side; and
//   - the BACKEND facts — runtime library load and model presence — via the
//     `fa_preflight` Tauri command (`src-tauri/src/fa_preflight.rs`).
//
// It answers, before an Apply Sync commits to inference, the question the
// FA-fallback entry used to answer only AFTER the run: is high-precision sync
// going to work, and if not, exactly what is missing. `App.tsx` folds the result
// into one durable `fa-preflight` sync-log entry the user sees up front.
//
// NEVER THROWS. Like `runForcedAlignmentForSync`, every failure — a rejected
// IPC call, a missing runtime, an unresolved language — becomes a structured
// "not ready" result, not an exception. A pre-flight that crashed the sync it
// was meant to de-risk would be worse than no pre-flight.
// ---------------------------------------------------------------------------

import { invoke } from '@tauri-apps/api/core';
import { isFaCapable, resolveFaLanguage } from './faGate';
import { describeInvokeError } from './invokeError';
import { FA_SUPPORTED_LANGUAGES } from './forcedAlignmentRun';
import type { FaLanguageCode } from './faTextNormalize';
import type { Project } from '../types';

/** Mirrors `fa_preflight.rs`'s `FaPreflightReport` (serde camelCase). */
export interface FaPreflightReport {
  featureCompiled: boolean;
  runtimeOk: boolean;
  runtimeDetail: string;
  modelPresent: boolean;
  modelDetail: string;
  language: string;
}

/** The combined readiness verdict App.tsx logs and (optionally) acts on. */
export interface FaPreflightResult {
  /** True only when every gate is satisfied: capable, a supported language
   *  resolved, the model is present, and the runtime library loaded. When this
   *  is true, a subsequent `runForcedAlignmentForSync` should not fall back for
   *  a readiness reason. */
  ready: boolean;
  /** The language the gate resolved to (sticky choice, else `-l auto`
   *  detection), or undefined when the project has neither. */
  resolvedLanguage: string | undefined;
  /** Individual signals, each independently loggable. `undefined` where the
   *  check did not run (e.g. backend checks are skipped when not capable). */
  capable: boolean;
  languageSupported: boolean;
  featureCompiled?: boolean;
  runtimeOk?: boolean;
  runtimeDetail?: string;
  modelPresent?: boolean;
  modelDetail?: string;
  /** One-line human summary for the log message. */
  summary: string;
  /** The first BLOCKING cause's verbatim detail (runtime/model text or the
   *  language problem), or undefined when ready. */
  blockingDetail?: string;
  /** The action for the user when not ready, or undefined when ready. */
  fixHint?: string;
}

/**
 * Runs the FA readiness pre-flight for `project`. Assumes the caller only
 * invokes it when the FA gate is OPEN (capability + the per-project switch) —
 * it still re-checks capability defensively so a direct caller cannot get a
 * misleading "ready" out of a non-Tauri runtime.
 */
export async function runFaPreflight(
  project: Pick<Project, 'language' | 'detectedLanguage'> | null | undefined,
): Promise<FaPreflightResult> {
  const capable = isFaCapable();
  const resolvedLanguage = resolveFaLanguage(project);
  const languageSupported =
    resolvedLanguage !== undefined &&
    FA_SUPPORTED_LANGUAGES.includes(resolvedLanguage as FaLanguageCode);

  if (!capable) {
    return {
      ready: false,
      resolvedLanguage,
      capable,
      languageSupported,
      summary: 'FA pre-flight: not available — the desktop runtime (Tauri) is not present.',
      blockingDetail: 'forced alignment requires the desktop app; it cannot run in a plain browser.',
      fixHint: 'Run the desktop app (npm run tauri:dev / the built app) for high-precision sync.',
    };
  }

  if (!languageSupported) {
    const langText = resolvedLanguage === undefined ? 'none detected or set' : `"${resolvedLanguage}"`;
    return {
      ready: false,
      resolvedLanguage,
      capable,
      languageSupported,
      summary: `FA pre-flight: not ready — no forced-alignment model for the project language (${langText}).`,
      blockingDetail:
        resolvedLanguage === undefined
          ? 'no language is set and none was detected — transcribe the voiceover first, or set the language in Project Settings.'
          : `${langText} is outside the five FA-supported languages (${FA_SUPPORTED_LANGUAGES.join(', ')}).`,
      fixHint:
        'Set the project language to English, Spanish, French, Portuguese, or German in Project Settings, or leave high-precision sync off for this project.',
    };
  }

  // Language resolves and is supported — ask the backend about the model + the
  // native runtime.
  let report: FaPreflightReport;
  try {
    report = await invoke<FaPreflightReport>('fa_preflight', { language: resolvedLanguage });
  } catch (err) {
    const detail = describeInvokeError(err);
    return {
      ready: false,
      resolvedLanguage,
      capable,
      languageSupported,
      summary: 'FA pre-flight: could not be checked — the readiness probe was rejected.',
      blockingDetail: detail,
      fixHint: 'Re-run Apply Sync. If it keeps happening, restart the app.',
    };
  }

  const ready = report.featureCompiled && report.runtimeOk && report.modelPresent;
  if (ready) {
    return {
      ready: true,
      resolvedLanguage,
      capable,
      languageSupported,
      featureCompiled: report.featureCompiled,
      runtimeOk: report.runtimeOk,
      runtimeDetail: report.runtimeDetail,
      modelPresent: report.modelPresent,
      modelDetail: report.modelDetail,
      summary: `FA pre-flight: ready — runtime loaded and model present for "${resolvedLanguage}".`,
    };
  }

  // Not ready: name the FIRST blocking cause, in the order the real run would
  // hit it (feature → runtime → model).
  let blockingDetail: string;
  let fixHint: string;
  if (!report.featureCompiled) {
    blockingDetail = report.runtimeDetail;
    fixHint = 'This build was compiled without forced alignment. Use a tauri:dev:fa / fa-inference build.';
  } else if (!report.runtimeOk) {
    blockingDetail = report.runtimeDetail;
    fixHint = 'The onnxruntime library could not load — re-provision it per src-tauri/onnxruntime/README.md.';
  } else {
    blockingDetail = report.modelDetail;
    fixHint = `Install the forced-alignment model for "${resolvedLanguage}" (see the searched paths above), then run Apply Sync again.`;
  }

  return {
    ready: false,
    resolvedLanguage,
    capable,
    languageSupported,
    featureCompiled: report.featureCompiled,
    runtimeOk: report.runtimeOk,
    runtimeDetail: report.runtimeDetail,
    modelPresent: report.modelPresent,
    modelDetail: report.modelDetail,
    summary: `FA pre-flight: not ready — ${!report.runtimeOk ? 'the alignment runtime did not load' : 'the alignment model is missing'} for "${resolvedLanguage}".`,
    blockingDetail,
    fixHint,
  };
}

// ---------------------------------------------------------------------------
// WS2 T4.1 Step 3 — the raw, per-language probe, for UI that needs the backend
// facts WITHOUT the project-shaped verdict `runFaPreflight` computes.
//
// WHY PROJECT SETTINGS' PACK DETECTOR CANNOT USE `isFaCapable()` ALONE.
// `faGate.ts`'s `isFaCapable()` is `isTauri()` and nothing more — it answers
// "is the IPC bridge present", which is necessary and nowhere near sufficient.
// `fa-inference` is NOT in `Cargo.toml`'s default feature set, so in a plain
// `tauri:dev`/`tauri:build` binary the bridge is present, `isFaCapable()`
// returns true, and `fa_align` returns `NotImplemented` for every run
// (`src-tauri/src/fa.rs`'s `#[cfg(not(feature = "fa-inference"))]` arm). A
// detector built on `isFaCapable()` would therefore report an installed pack
// as USABLE in the exact binary that ships today, which is a worse lie than
// reporting nothing.
//
// NO NEW PROBE WAS NEEDED, and no `not_implemented` round-trip either.
// `fa_preflight` already returns `featureCompiled` straight from a
// `#[cfg(feature = "fa-inference")]`, so it reports the BUILD FACT directly
// rather than inferring it from a failed alignment. It is cheap by
// construction (a path stat plus a dlopen + ort env init; it explicitly does
// NOT hash the ~1.2 GiB model) because it was designed to run before every FA
// sync. This is that same command, called with an explicit language instead of
// one resolved from a project.
// ---------------------------------------------------------------------------

/**
 * Runs the backend readiness probe for one language. Returns `null` — never
 * throws — when the runtime is not Tauri-capable at all or the IPC call is
 * rejected; callers render "unknown", which is honestly distinct from both
 * "ready" and "the pack is missing".
 */
export async function probeFaReadiness(language: string): Promise<FaPreflightReport | null> {
  if (!isFaCapable()) return null;
  try {
    return await invoke<FaPreflightReport>('fa_preflight', { language });
  } catch {
    return null;
  }
}
